### ISSUE TITLE
Cilium pods fail to start after upgrade from 1.15.3 to 1.15.4

### ISSUE THREAD

**Issue by roosmaa** [cilium/cilium#31944]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Started a minor version upgrade (1.15.3 to 1.15.4) and the pod started failing with error:

```
failed to start: daemon creation failed: error while initializing daemon: failed while reinitializing datapath: listing routing rules: address family not supported by protocol
```

### Cilium Version

1.15.4

### Kernel Version

Linux version 6.1.82-talos (@buildkitsandbox) (gcc (GCC) 13.2.0, GNU ld (GNU Binutils) 2.41) #1 SMP Tue Mar 19 17:48:01 UTC 2024

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"29", GitVersion:"v1.29.2", GitCommit:"4b8e819355d791d96b7e9d9efe4cbafae2311c88", GitTreeState:"clean", BuildDate:"2024-02-14T10:32:40Z", GoVersion:"go1.21.7", Compiler:"gc", Platform:"linux/amd64"}

### Regression

1.15.3

### Relevant log output

```shell
time="2024-04-13T07:38:47Z" level=info msg="Endpoints restored" failed=4 restored=23 subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="Addressing information:" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  Cluster-Name: default" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  Cluster-ID: 0" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  Local node-name: abm-03" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  Node-IPv6: <nil>" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  External-Node IPv4: 10.22.1.3" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  Internal-Node IPv4: 10.26.2.34" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  IPv4 allocation prefix: 10.26.2.0/24" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  IPv4 native routing prefix: 10.26.0.0/16" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  Loopback IPv4: 169.254.42.1" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  Local IPv4 addresses:" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  - 10.22.1.3" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="  - 10.26.2.34" subsys=daemon
time="2024-04-13T07:38:47Z" level=info msg="Node updated" clusterName=default nodeName=abm-03 subsys=nodemanager
time="2024-04-13T07:38:47Z" level=info msg="Adding local node to cluster" node=abm-03 subsys=nodediscovery
time="2024-04-13T07:38:47Z" level=info msg="Node updated" clusterName=default nodeName=abm-02 subsys=nodemanager
time="2024-04-13T07:38:47Z" level=info msg="Creating or updating CiliumNode resource" node=abm-03 subsys=nodediscovery
time="2024-04-13T07:38:47Z" level=info msg="Waiting until all pre-existing resources have been received" subsys=k8s-watcher
time="2024-04-13T07:38:47Z" level=info msg="Initializing identity allocator" subsys=identity-cache
time="2024-04-13T07:38:47Z" level=info msg="Allocating identities between range" cluster-id=0 max=65535 min=256 subsys=identity-cache
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_host.forwarding sysParamValue=1
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_host.rp_filter sysParamValue=0
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_host.accept_local sysParamValue=1
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_host.send_redirects sysParamValue=0
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_net.forwarding sysParamValue=1
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_net.rp_filter sysParamValue=0
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_net.accept_local sysParamValue=1
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_net.send_redirects sysParamValue=0
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.core.bpf_jit_enable sysParamValue=1
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.all.rp_filter sysParamValue=0
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.fib_multipath_use_neigh sysParamValue=1
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=kernel.unprivileged_bpf_disabled sysParamValue=1
time="2024-04-13T07:38:47Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=kernel.timer_migration sysParamValue=0
time="2024-04-13T07:38:47Z" level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
time="2024-04-13T07:38:48Z" level=info msg="Updated link /sys/fs/bpf/cilium/socketlb/links/cgroup/cil_sock4_recvmsg for program cil_sock4_recvmsg" subsys=socketlb
time="2024-04-13T07:38:48Z" level=info msg="Updated link /sys/fs/bpf/cilium/socketlb/links/cgroup/cil_sock4_connect for program cil_sock4_connect" subsys=socketlb
time="2024-04-13T07:38:48Z" level=info msg="Updated link /sys/fs/bpf/cilium/socketlb/links/cgroup/cil_sock4_getpeername for program cil_sock4_getpeername" subsys=socketlb
time="2024-04-13T07:38:48Z" level=info msg="Updated link /sys/fs/bpf/cilium/socketlb/links/cgroup/cil_sock4_post_bind for program cil_sock4_post_bind" subsys=socketlb
time="2024-04-13T07:38:48Z" level=info msg="Updated link /sys/fs/bpf/cilium/socketlb/links/cgroup/cil_sock4_sendmsg for program cil_sock4_sendmsg" subsys=socketlb
time="2024-04-13T07:38:48Z" level=info msg="Iptables rules installed" subsys=iptables
time="2024-04-13T07:38:48Z" level=error msg="Start hook failed" error="daemon creation failed: error while initializing daemon: failed while reinitializing datapath: listing routing rules: address family not supported by protocol" function="cmd.newDaemonPromise.func1 (cmd/daemon_main.go:1686)" subsys=hive
time="2024-04-13T07:38:48Z" level=info msg=Stopping subsys=hive
[...snip...]
time="2024-04-13T07:38:48Z" level=info msg="Stop hook executed" duration="68.626µs" function="gops.registerGopsHooks.func2 (pkg/gops/cell.go:50)" subsys=hive
time="2024-04-13T07:38:48Z" level=fatal msg="failed to start: daemon creation failed: error while initializing daemon: failed while reinitializing datapath: listing routing rules: address family not supported by protocol" subsys=daemon
time="2024-04-13T07:38:48Z" level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
```

### Anything else?

I have IPv6 disabled on kernel level on my nodes, considering the error it could be related.

Looking at the error message and reading the code, it's failing somewhere in this block https://github.com/cilium/cilium/blob/v1.15.4/pkg/datapath/loader/base.go#L478-L486 , I think. But I cannot track down the `listing routing rules` part of the error from there on out.

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by roosmaa** [comment]
Disabling the L7 proxy in the Helm values.yaml worksaround the issue.

```
l7Proxy: false
```

**Comment by nico151999** [comment]
@roosmaa as long as one does not need the l7proxy, yes. Otherwise, this is not an option (e.g. if one uses the Gateway API).

**Comment by julianwiedmann** [comment]
yep, maybe https://github.com/cilium/cilium/pull/31777 needs to check for `errors.Is(err, syscall.EAFNOSUPPORT)` a bit earlier and just exit gracefully.

(@jschwinger233 ^^)

**Comment by julianwiedmann** [comment]
Should be fixed with `v1.15.5` by https://github.com/cilium/cilium/pull/32007, please re-open if that's not the case :).
